pkg/controller/externalippool/controller_test.go

```diff
@@ -0,0 +1,718 @@
+	assert.True(t, changed)
+	allocator, exists = controller.getIPAllocator(ipPool.Name)
+	require.True(t, exists)
+	assert.Equal(t, 3, len(allocator))
+	assert.Equal(t, 14, allocator.Total())
+
+	// IPv6 CIDR shouldn't exclude broadcast address, so total should be increased by 15.
+	ipPool.Spec.IPRanges = append(ipPool.Spec.IPRanges, antreacrds.IPRange{CIDR: "2021:3::aaa1/124"})
+	changed = controller.createOrUpdateIPAllocator(ipPool)
+	assert.True(t, changed)
+	allocator, exists = controller.getIPAllocator(ipPool.Name)
+	require.True(t, exists)
+	assert.Equal(t, 4, len(allocator))
+	assert.Equal(t, 29, allocator.Total())
+
+	// When there is no change, the method should do nothing and the return value should be false.
+	changed = controller.createOrUpdateIPAllocator(ipPool)
+	assert.False(t, changed)
+	allocator, exists = controller.getIPAllocator(ipPool.Name)
+	require.True(t, exists)
+	assert.Equal(t, 4, len(allocator))
+	assert.Equal(t, 29, allocator.Total())
+}
+
+func TestIPPoolEvents(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	context, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	controller := newController(nil)
+	consumerCh := make(chan string)
+	controller.AddEventHandler(
+		func(ippool string) {
+			consumerCh <- ippool
+		})
+	controller.crdInformerFactory.Start(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+	go controller.Run(stopCh)
+	require.True(t, cache.WaitForCacheSync(stopCh, controller.HasSynced))
+	// ADD event
+	eip, err := controller.crdClient.CrdV1alpha2().ExternalIPPools().Create(context,
+		newExternalIPPool("eip1", "", "10.10.10.2", "10.10.10.3"),
+		metav1.CreateOptions{},
+	)
+	require.NoError(t, err)
+	assert.Equal(t, "eip1", <-consumerCh)
+	// UPDATE event
+	eip.Spec.IPRanges[0].End = "10.10.10.4"
+	eip, err = controller.crdClient.CrdV1alpha2().ExternalIPPools().Update(context,
+		eip,
+		metav1.UpdateOptions{},
+	)
+	require.NoError(t, err)
+	assert.Equal(t, "eip1", <-consumerCh)
+	// DELETE event
+	err = controller.crdClient.CrdV1alpha2().ExternalIPPools().Delete(context,
+		eip.Name,
+		metav1.DeleteOptions{},
+	)
+	require.NoError(t, err)
+	assert.Equal(t, "eip1", <-consumerCh)
+}
+func TestConsumersRestoreIPAllocation(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	eip := newExternalIPPool("eip1", "", "10.10.10.2", "10.10.10.10")
+	controller := newController([]runtime.Object{eip})
+	controller.AddEventHandler(func(ippool string) {})
+	controller.AddEventHandler(func(ippool string) {})
+	controller.crdInformerFactory.Start(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+	go controller.Run(stopCh)
+	require.True(t, cache.WaitForCacheSync(stopCh, controller.HasSynced))
+	consumerCh := make(chan struct{})
+	consumerEndCh := make(chan struct{})
+	go func() {
+		require.NoError(t, controller.UpdateIPAllocation("eip1", net.ParseIP("10.10.10.2")))
+		controller.RestoreIPAllocations(nil)
```

Comment:
I think the IP allocation should be restored via `RestoreIPAllocations` instead of `UpdateIPAllocation`?

Reply:
refactored this test case.